---		
layout: default		
title: 'Gene-set analysis'		
---		
		
# Gene-set analysis

### Introduction and data

The follwing packages are used in this tutorial: `DESeq2`, `biomaRt`, `piano`, `snow`, `snowfall`. In case you haven't installed them yet it could be convenient to do so before starting (you can potentially skip `snow` and `snowfall`). We will perform gene-set analysis on the output from the tutorial on Differential expression analysis of RNA-seq data using DESeq. A quick recap of the essential code for the differential expression analysis is included below, in case you did not save the output from that analysis:

``` {r run DESeq2}
library(DESeq2)
counts <- read.delim("count_table.txt")
samples <- data.frame(timepoint = rep(c("ctrl", "t2h", "t6h", "t24h"), each=3))
ds <- DESeqDataSetFromMatrix(countData=counts, colData=samples, design=~timepoint)
colnames(ds) <- colnames(counts)
ds <- DESeq(ds)
res <- results(ds, c("timepoint","t24h","ctrl"))
res <- res[ ! (is.na(res$pvalue) | is.na(res$padj)), ] 
# Here we also exclude genes with adjusted p-values = NA
```

Let's save the information we need (fold-changes and adjusted p-values):

``` {r save info}
geneLevelStats <- as.data.frame(res[,c("log2FoldChange","padj")])

```

It will be handy to have the gene names along with the Ensembl IDs, so let's fetch them:

```{r load better Names}
library(biomaRt) # Install the biomaRt package (Bioconductor) if this command does not work

# Get the Ensembl ID to gene name mapping:
mart <- useMart("ENSEMBL_MART_ENSEMBL", dataset = "hsapiens_gene_ensembl", host="www.ensembl.org")
ensembl2name <- getBM(attributes=c("ensembl_gene_id","external_gene_name"),mart=mart)

# Merge it with our gene-level statistics:
geneLevelStats <- merge(x=geneLevelStats, y=ensembl2name, by.x=0, by.y=1, all.x=TRUE)
colnames(geneLevelStats) <- c("ensembl","log2fc","padj","gene")
rownames(geneLevelStats) <- geneLevelStats[,1]

# Sort the gene-level statistics by adjusted p-value:
geneLevelStats <- geneLevelStats[order(geneLevelStats$padj),]
head(geneLevelStats)

```

### Gene-set analysis

Looking at only the top 100 genes (or genes with a adjusted p-value below some cutoff) has the drawback of excluding a lot of information. Gene-set analysis (GSA) takes into account the "scores" (which can be e.g. p-values, fold-changes, etc) of all genes. This allows us to also detect small but coordinate changes converging on specific biological functions or other themes. In this tutorial we will use the piano package to perform GSA:

```{r run piano}
#source("http://bioconductor.org/biocLite.R")
#biocLite("piano")


library(piano) # Install the piano package (Bioconductor) if this command does not work
```

First we need to construct our gene-set collection, we will be looking at so called Hallmark gene-sets from the MSigDB in this example (see this [paper](http://www.cell.com/cell-systems/abstract/S2405-4712(15)00218-5)). Download the Hallmark gene-set collection from [here](http://software.broadinstitute.org/gsea/msigdb/download_file.jsp?filePath=/resources/msigdb/5.1/h.all.v5.1.symbols.gmt). Note that you need to sign up with an email adress to gain access. Visit [this link](http://software.broadinstitute.org/gsea/msigdb/collections.jsp#H) if the first link does not work.

```{r load geneset}
# Load the gene-set collection into piano format:
gsc <- loadGSC("h.all.v5.1.symbols.gmt", type="gmt") # Check that the filename matches the file that you downloaded
gsc # Always take a look at the GSC object to see that it loaded correctly
```

Now we are ready to run the GSA:

```{r  run GSA}
#library(snowfall); library(snow) # Install snow and snowfall (CRAN) if you want to 
# run on multiple cores, otherwise omit the ncpus argument below in the call to runGSA
padj <- geneLevelStats$padj
log2fc <- geneLevelStats$log2fc
names(padj) <- names(log2fc) <- geneLevelStats$gene
gsaRes <- runGSA(padj, log2fc, gsc=gsc)
```

 
The runGSA function uses the adjusted p-values to score the genes and the log2-foldchange for information about the direction of change.

We can visualize the results in different ways, for instance using a network plot showing the significant gene-sets and the overlap of genes between sets:

```{r pathwayAnalysis}
networkPlot(gsaRes, "distinct", "both", adjusted=T, ncharLabel=Inf, significance=0.01,
            nodeSize=c(3,20), edgeWidth=c(1,5), overlap=10, 
            scoreColors=c("red", "orange", "yellow", "blue", "lightblue", "lightgreen"))
```


The function `GSAsummaryTable` can be used to export the complete results.

```{r view table}
GSAsummaryTable(gsaRes)
# if you want to you can also save this as a file:
# GSAsummaryTable(gsaRes, save=T, file="gsares.txt")
```


The `geneSetSummary` function can be used to explore specific gene-sets in more detail.


For instance, we can make a boxplot of the -log10(adjusted p-values) of the genes in the gene-set HALLMARK\_DNA\_REPAIR and compare that to the distribution of all genes:

```{r boxplot}
boxplot(list(-log10(geneLevelStats$padj),
             -log10(geneSetSummary(gsaRes,"HALLMARK_DNA_REPAIR")$geneLevelStats)),
        names=c("all","HALLMARK_DNA_REPAIR"))
```


From here, you can dig in to the results on the gene-set level further and start making hypothesis of what is happening with the biology behind your data. You can also try to run GSA with other gene-set collections (e.g. from [MSigDB](http://software.broadinstitute.org/gsea/msigdb/)) or using another GSA method (see `?runGSA`).

### Further reading

-   Piano webpage, with more information and link to publication: <http://www.sysbio.se/piano>
-   GSEA paper: <http://www.pnas.org/content/102/43/15545.full>
-   A couple of reviews:
-   <http://bib.oxfordjournals.org/content/9/3/189.full>
-   <http://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-10-47>
